### PR TITLE
Advanced command gestion

### PR DESCRIPTION
**Pull request by SimonCROS**
```diff
PRIVMSG
*** :You have not registered
PRIVMSG Someone ?
*** :You have not registered
USER scros 0 * :Simon Cros
USER adel-sol 0 * :Arsène Del Sol
*** :Unauthorized command (already registered)
NICK Saumon
PRIVMSG Someone ?
*** Someone :No such nick/channel
PRIVMSG Lucasquette
*** :No text to send
PRIVMSG
*** :No recipient given (PRIVMSG)
PRIVMSG Lucasquette :Hello World !!!
```
Messages are added manually by me, for now only the response codes are displayed.

#4 and #5 started, need to check the format before closing.

### CODE REVIEW

File: includes/modules/api/command/Command.hpp [hunk 1]
```diff
@@ -1,16 +1,15 @@
...
 
 class Command {
 private:
 	friend class CommandSpec;
 
 	string _name;
-	map<string, void *const>& _args;
+	map<string, void *const> _args;
 
 	Command(const string& name, map<string, void *const>& args) : _name(name), _args(args) {}
```
Comment (MauriceNDS) [inline includes/modules/api/command/Command.hpp:15]:
```c++
_args.at();
```
only in C++ 11

File: includes/modules/api/User.hpp [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+#include "ft_irc.hpp"
+
+#include "api/Client.hpp"
+
+class User : public Client {
+private:
+	string nickname;
+	string username;
+	string realname;
+	bool oper;
```
Comment (MauriceNDS) [inline includes/modules/api/User.hpp]: Remove operator functions.

File: includes/modules/api/Channel.hpp [hunk 3]
```diff
@@ -0,0 +1,35 @@
...
+#include "api/CommandSender.hpp"
+#include "api/User.hpp"
+
+class Channel : public CommandSender {
+private:
+	string name;
+    vector<User *> users;
+
+public:
+	Channel(string name) : name(name) {}
```
Comment (MauriceNDS) [inline includes/modules/api/Channel.hpp]:
```suggestion
	Channel(const string& name) : name(name) {}
```

File: includes/modules/api/command/CommandSpec.hpp [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+public:
+	class Builder {
+	private:
+		string _name;
+		vector<pair<string, CommandElement *> > _parameters;
+		CommandExecutor *_executor;
+		Middleware *_middleware;
+
+	public:
+		Builder() : _executor(nullptr), _middleware(nullptr) {}
```
Comment (MauriceNDS) [inline includes/modules/api/command/CommandSpec.hpp]:
```suggestion
		Builder() : _executor(NULL), _middleware(NULL) {}
```
nullptr is in c++ 11

File: includes/modules/api/command/GenericArguments.hpp [hunk 5]
```diff
@@ -38,18 +40,18 @@ class OptionalCommandElement : public CommandElement {
...
+	bool isRequired() const {
+		return false;
 	}
 
-	void *parseValue(const string& arg) const {
+	void *parseValue(const string& arg, MessageEvent& event) const {
 		if (arg.empty())
 			return new const void*(nullptr);
-		return new const void*(subtype->parseValue(arg));
+		return new const void*(subtype->parseValue(arg, event));
```
Comment (MauriceNDS) [inline includes/modules/api/command/GenericArguments.hpp]: If subtype conversion fails, return NULL.

File: includes/modules/api/command/GenericArguments.hpp [hunk 6]
```diff
@@ -67,21 +69,18 @@ class ListCommandElement : public CommandElement {
...
+		std::istringstream ss(arg);
 		string temp;
-		while (ss >> temp)
+		while (std::getline(ss, temp, ','))
 			tokens.push_back(temp);
-		vector<string>::iterator tokens_it;
 
 		for (tokens_it = tokens.begin(); tokens_it != tokens.end(); tokens_it++) {
-			val->push_back(static_cast<T*>(subtype->parseValue(*tokens_it)));
+			val->push_back(static_cast<T*>(subtype->parseValue(*tokens_it, event)));
```
Comment (MauriceNDS) [inline includes/modules/api/command/GenericArguments.hpp]: If one iteration fails, continue but return NULL and free `val` and all the other elements.

File: includes/modules/core/command/NickCommand.hpp [hunk 7]
```diff
@@ -0,0 +1,29 @@
...
+#include "server/Response.hpp"
+
+class NickCommand : public CommandExecutor {
+
+	Response execute(const Command& cmd, CommandSender& sender) {
+		User& user = dynamic_cast<User&>(sender);
+
+		string nickname = cmd.getArg<string>("nickname");
+		if (nickname.empty())
+			return ERR_NONICKNAMEGIVEN;
```
Comment (MauriceNDS) [inline includes/modules/core/command/NickCommand.hpp]: Cannot be empty, set this error during CommandSpec creation.

**Review by MauriceNDS — CHANGES_REQUESTED** [review]

File: includes/modules/core/command/NickCommand.hpp (hunk 7 above)
Comment (SimonCROS) [reply includes/modules/core/command/NickCommand.hpp]: Can be empty if the command is `NICK :`, but the response may be `ERR_ERRONEUSNICKNAME`.
`ERR_NONICKNAMEGIVEN` may be send when the command is `NICK`.

**Review by Owllyy — APPROVED** [review]

**Review by MauriceNDS — APPROVED** [review]

**Pull request merged**